### ISSUE TITLE
[SalesForce Case] 00695779 Unable to get past 401 error with kubernetes authenticator

### ISSUE THREAD

**Issue by garkler** [cyberark/conjur-authn-k8s-client#21]
[SaleForce Case](https://cyberark.my.salesforce.com/5002J00001Ezprr)

**From the Customer:**
OCP 3.9 env 
Follower deployed in OCP
Implementation on customer site

The authenticator is deployed as a sidecar. The service account is oc-test-app-summon-sidecar. The application deployment config has the following env: 

env: 
```
- name: CONJUR_VERSION 
value: '5' 
- name: CONJUR_APPLIANCE_URL 
value: 'https://conjur-follower.capgroup.com' 
- name: CONJUR_AUTHN_URL 
value: 'https://conjur-follower.capgroup.com/api/authn-k8s/dev' 
- name: CONJUR_ACCOUNT 
value: prod 
- name: CONJUR_AUTHN_LOGIN 
value: host/conjur/authn-k8s/dev/apps/bb000024-conjur-dev-01/*/* 
- name: CONJUR_SSL_CERTIFICATE 
valueFrom: 
configMapKeyRef: 
key: ssl-certificate 
name: conjur-cert 
volumeMounts: 
- mountPath: /run/conjur 
name: conjur-access-token 
```

The sidecar env: 
```
- image: cgdevregistry.capgroup.com/cgbuild/cgsec/cgconjur-authn-k8s-client 
imagePullPolicy: Always 
name: authenticator 
env: 
- name: CONTAINER_MODE 
value: sidecar 
- name: MY_POD_NAME 
valueFrom: 
fieldRef: 
fieldPath: metadata.name 
- name: MY_POD_NAMESPACE 
valueFrom: 
fieldRef: 
fieldPath: metadata.namespace 
- name: MY_POD_IP 
valueFrom: 
fieldRef: 
fieldPath: status.podIP 
- name: CONJUR_VERSION 
value: '5' 
- name: CONJUR_APPLIANCE_URL 
value: 'https://conjur-follower.capgroup.com' 
- name: CONJUR_AUTHN_URL 
value: 'https://conjur-follower.capgroup.com/api/authn-k8s/dev' 
- name: CONJUR_ACCOUNT 
value: prod 
- name: CONJUR_AUTHN_LOGIN 
value: host/conjur/authn-k8s/dev/apps/bb000024-conjur-dev-01/*/* 
- name: CONJUR_SSL_CERTIFICATE 
valueFrom: 
configMapKeyRef: 
key: ssl-certificate 
name: conjur-cert 
volumeMounts: 
- mountPath: /run/conjur 
name: conjur-access-token 
imagePullSecrets: 
- name: dockerpullsecret 
volumes: 
- name: conjur-access-token 
emptyDir: 
medium: Memory 
```

The policy to create the host: 
--- 
```
# This policy defines a layer of whitelisted identities permitted to authenticate to the authn-k8s endpoint. 
- !policy 
id: conjur/authn-k8s/dev/apps 
annotations: 
description: Identities permitted to authenticate 
body: 
- !layer 
annotations: 
description: Layer of authenticator identities permitted to call authn svc 
- &hosts 
- !host 
id: bb00000024-conjur-dev-01/*/* 
annotations: 
kubernetes/authentication-container-name: authenticator 
openshift: "true" 
- !host 
id: bb00000024-conjur-dev-01/service_account/oc-test-app-summon-sidecar 
annotations: 
kubernetes/authentication-container-name: authenticator 
openshift: "true" 
- !grant 
role: !layer 
members: *hosts 
```

The sidecar logs just show 401 authentication failure errors and I don't know what else to check.

**Comment by jtuttle** [comment]
The problems we encountered during this call were:

1. The app container was configured to authenticate with a service account identity but was being deployed using the namespace's default service account. This would be flagged during validation in Conjur and cause an authentication failure. We addressed this by updating the application manifest and redeploying the pod.
1. The follower pod was being deployed with the namespace's default service account instead of a `conjur-authenticator` service account that has been granted all the necessary RBAC privileges to perform pod inspection and certificate delivery. This would prevent Conjur from validating the application pod and injecting a certificate. We addressed this by creating such a service account, loading role and role-bindings for it, updating the follower manifest, and redeploying the follower pod.
1. The new service account had not been granted permission to deploy Docker containers with the root user. We addressed this by having a cluster admin run:
```
oc adm policy add-scc-to-user anyuid system:serviceaccount:bb00000024-conjur-dev-01:conjur-authenticator
```
